MatrixSDK/Crypto/Dehydration/MXDehydrationService.m

```diff
@@ -0,0 +1,260 @@
+        failure(error);
+    }];
+}
+
+- (void)rehydrateDeviceWithSuccess:(void (^)(void))success
+                           failure:(void (^)(NSError *error))failure
+{
+    if (session.crypto)
+    {
+        MXLogError(@"[MXDehydrationManager] rehydrateDevice: Cannot rehydrate device after crypto is initialized.");
+        failure([NSError errorWithDomain:kMXNSErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: @"Cannot rehydrate device after crypto is initialized"}]);
+        return;
+    }
+    
+    
+    MXKeyData * keyData = [[MXKeyProvider sharedInstance] requestKeyForDataOfType:MXDehydrationServiceKeyDataType isMandatory:NO expectedKeyType:kRawData];
+    if (!keyData)
+    {
+        MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: No dehydrated key.");
+        success();
+        return;
+    }
+    NSData *key = ((MXRawDataKey*) keyData).key;
+    
+    MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: getting dehydrated device.");
+    [self->session.matrixRestClient dehydratedDeviceWithSuccess:^(MXDehydratedDevice *device) {
+        if (!device || !device.deviceId)
+        {
+            MXLogDebug(@"[MXSession] rehydrateDevice: No dehydrated device found.");
+            success();
+            return;
+        }
+        
+        if (![device.algorithm isEqual:MXDehydrationAlgorithm])
+        {
+            MXLogError(@"[MXDehydrationManager] rehydrateDevice: Wrong algorithm for dehydrated device.");
+            failure([NSError errorWithDomain:kMXNSErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: @"Wrong algorithm for dehydrated device"}]);
+            return;
+        }
+        
+        NSError *error = nil;
+        MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: unpickling dehydrated device.");
+        OLMAccount *account = [[OLMAccount alloc] initWithSerializedData:device.account key:key error:&error];
+        
+        MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: account deserialized %@", account.identityKeys);
+
+        if (error)
+        {
+            MXLogError(@"[MXDehydrationManager] rehydrateDevice: Failed to unpickle device account with error: %@.", error);
+            failure([NSError errorWithDomain:kMXNSErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: @"Failed to unpickle device account"}]);
+            return;
+        }
+        
+        MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: device unpickled %@", account);
+        
+        [self->session.matrixRestClient claimDehydratedDeviceWithId:device.deviceId Success:^(BOOL isClaimed) {
+            if (!isClaimed)
+            {
+                MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: device already claimed.");
+                success();
+                return;
+            }
+
+            MXLogDebug(@"[MXDehydrationManager] rehydrateDevice: using dehydrated device");
+            self->session.matrixRestClient.credentials.deviceId = device.deviceId;
```

Comment:
To go further on this architectural change, we could have some stateless static methods in MXCrypto that apply the change for the rehydrated device.
Thus, the next instantiated MXSession will start with the right data.
It seems safer than calling this method at the right time where right time window is quite small.

Reply:
MXDehydrationService can do those changes internally for the SDK user and just return the device id. But it means that no MXSession must run and this service should not depend on MXSession for that part.

The flow would be:
- Have a MXRestClient
- Init MXDehydrationService with it
- MXDehydrationService.rehydrate(). It will return a device id
- Let the SDK user store the device id
- Start a new MXRestClient with new credentials
- Init MXSession with it as usual. There should be no change on this existing code for this call